src/core/server/saved_objects/types.ts

```diff
@@ -277,6 +282,12 @@ export interface SavedObjectsType<Attributes = any> {
    * If defined, will be used to convert the type to an alias.
    */
   convertToAliasScript?: string;
+  /**
+   * If defined, allows a type to run a search query and return a query filter that may match any documents which may
+   * be excluded from the next migration upgrade process. Useful for cleaning up old documents which are no longer
+   * needed.
+   */
+  deleteOnUpgrade?: SavedObjectTypeExcludeFromUpgradeFilterHook;
```

Comment:
Nice! We'll need to be very explicit in guiding plugin owners on when to use a migration vs `deleteOnUpgrade` though. I'm assuming that migrations are registered for saved object types (and therefor all documents for those saved objects) and that `deleteOnUpgrade` is only to clean up a subset of these. We'll need to make sure saved object type owners know when to use what.

Reply:
Added to the list for follow ups to be done after we get this blocker merged.